commit 5aa66544807176c8c71906b7d2b65618c83ae19b (refs/changes/16/5016/1)
Author: Mike Gerdts <mike.gerdts@joyent.com>
Date:   2018-10-31T15:22:40+00:00 (11 months ago)
    
    OS-7333 expander.js could prevent baldness by sanity check non-string zone.attr.*

diff --git a/src/vm/node_modules/expander.js b/src/vm/node_modules/expander.js
index d79db38e..9b90ce13 100644
--- a/src/vm/node_modules/expander.js
+++ b/src/vm/node_modules/expander.js
@@ -135,6 +135,17 @@ Object.keys(proptable.properties).forEach(function (prop) {
         if (!property.hasOwnProperty('ignore')) {
             vmobj_zonexml_fields.push(prop);
         }
+        // zone.attr.* are stored as strings. If the type is not a string and
+        // there is no translator, save the developer a few hours of debugging.
+        if (property.zonexml.substring(0, 10) === 'zone.attr.'
+            && !property.loadValueTranslator
+            && property.payload && property.payload.type) {
+            var str_types = ['flat-object', 'string', 'uuid'];
+            assert(str_types.indexOf(property.payload.type) !== -1,
+                'prop "' + prop + '" type "' + property.payload.type
+                + '" is not one of "' + str_types.join('", "')
+                + '" and has no loadValueTranslator.');
+        }
     } else if (property.zonexml_depends) {
         /*
          * zonexml_depends is used for things like transition_expire which
